plaidml/edsl/tests/edsl_test.cc

```diff
@@ -1659,41 +1659,237 @@ TEST_F(CppEdsl, Tan) {
+  checkExact(program, {data, indices, updates}, {expected});
+}
+
+TEST_F(CppEdsl, ScatterNDUpdateSlice) {
+  auto D = Placeholder(DType::FLOAT32, {8});
+  // The only difference with 'Scatter1DUpdateSlice' is the shape of indices.
+  auto I = Placeholder(DType::INT32, {4, 1});
+  auto U = Placeholder(DType::FLOAT32, {4});
+  auto O = scatter(D, I, U).update(ScatterUpdateMode::ND);
+  auto program = makeProgram("scatter", {D, I, U}, {O});
+
+  std::vector<float> data = {1, 1, 1, 1, 1, 1, 1, 1};
+  std::vector<int32_t> indices = {4, 3, 1, 7};
+  std::vector<float> updates = {9, 10, 11, 12};
+  std::vector<float> expected = {1, 11, 1, 10, 9, 1, 1, 12};
+  checkExact(program, {data, indices, updates}, {expected});
 }
 
 TEST_F(CppEdsl, Scatter3D) {
+  auto D = Placeholder(DType::FLOAT32, {4, 4, 4});
   auto I = Placeholder(DType::INT32, {2});
   auto U = Placeholder(DType::FLOAT32, {2, 4, 4});
-  auto S = Placeholder(DType::INT32, {4, 4, 4});
-  auto O = scatter(U, I, S);
-  auto program = makeProgram("scatter", {I, U, S}, {O});
+  auto O = scatter(D, I, U);
+  auto program = makeProgram("scatter", {D, I, U}, {O});
 
+  // Don't bother initialize 'data' while not updating.
+  // Only the shape of data is needed.
+  std::vector<float> data;
   std::vector<int32_t> indices = {0, 2};
   std::vector<float> updates = {
       5, 5, 5, 5, 6, 6, 6, 6,  //
       7, 7, 7, 7, 8, 8, 8, 8,  //
       5, 5, 5, 5, 6, 6, 6, 6,  //
       7, 7, 7, 7, 8, 8, 8, 8   //
   };
-  std::vector<int32_t> shape = {4, 4, 4};
   std::vector<float> expected = {
       5, 5, 5, 5, 6, 6, 6, 6, 7, 7, 7, 7, 8, 8, 8, 8,  //
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  //
       5, 5, 5, 5, 6, 6, 6, 6, 7, 7, 7, 7, 8, 8, 8, 8,  //
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0   //
   };
-  checkExact(program, {indices, updates, shape}, {expected});
+  checkExact(program, {data, indices, updates}, {expected});
+}
+
+TEST_F(CppEdsl, ScatterDup3D) {
+  auto D = Placeholder(DType::FLOAT32, {4, 4, 4});
+  auto I = Placeholder(DType::INT32, {2});
+  auto U = Placeholder(DType::FLOAT32, {2, 4, 4});
+  auto O = scatter(D, I, U);
+  auto program = makeProgram("scatter", {D, I, U}, {O});
+
+  // Don't bother initialize 'data' while not updating.
+  // Only the shape of data is needed.
+  std::vector<float> data;
+  // Duplicate indices.
+  std::vector<int32_t> indices = {2, 2};
+  std::vector<float> updates = {
+      5, 5, 5, 5, 6, 6, 6, 6,  //
+      7, 7, 7, 7, 8, 8, 8, 8,  //
+      5, 5, 5, 5, 6, 6, 6, 6,  //
+      7, 7, 7, 7, 8, 8, 8, 8   //
+  };
+  std::vector<float> expected = {
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  //
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  //
+      10, 10, 10, 10, 12, 12, 12, 12, 14, 14, 14, 14, 16, 16, 16, 16,  //
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0   //
+  };
+  checkExact(program, {data, indices, updates}, {expected});
+}
+
+TEST_F(CppEdsl, Scatter3DUpdateSlice) {
```

Comment:
Please rename to ScatterNDUpdateSlice

Reply:
My mistake.. I think there is another test with the same name... please keep it as Scatter3D..